pkg/intermediate/aggregate.go

```diff
@@ -0,0 +1,284 @@
+package intermediate
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/vmware/go-ipfix/pkg/entities"
+	"github.com/vmware/go-ipfix/pkg/registry"
+)
+
+type aggregation struct {
+	// tupleRecordMap maps each connection with its records
+	tupleRecordMap map[Tuple][]entities.Record
+	// tupleRecordLock allows multiple readers or one writer at the same time
+	tupleRecordLock sync.RWMutex
+	// workerPool is for storing worker channels to process the messages
+	workerPool chan chan *entities.Message
+	// messageChan is the channel to receive the message
+	messageChan chan *entities.Message
+	// workerNum is the number of workers to process the messages
+	workerNum int
+	// workerList is the list of workers
+	workerList []*worker
+	// correlateFields are the fields to be filled in correlating process
+	correlateFields []string
+}
+
+type Tuple struct {
+	SourceAddress      string
+	DestinationAddress string
+	Protocol           uint8
+	SourcePort         uint16
+	DestinationPort    uint16
+}
+
+func InitAggregationProcess(messageChan chan *entities.Message, workerNum int, correlateFields []string) (*aggregation, error) {
+	if messageChan == nil {
+		return nil, fmt.Errorf("Cannot create aggregation process without message channel.")
+	} else if workerNum <= 0 {
+		return nil, fmt.Errorf("Worker number cannot be <= 0.")
+	}
+	return &aggregation{
+		make(map[Tuple][]entities.Record),
+		sync.RWMutex{},
+		make(chan chan *entities.Message),
+		messageChan,
+		workerNum,
+		make([]*worker, 0),
+		correlateFields,
+	}, nil
+}
+
+func (a *aggregation) Start() {
+	for i := 0; i < a.workerNum; i++ {
+		w := createWorker(i, a.workerPool, a.AggregateMsgBy5Tuple)
+		w.start()
+		a.workerList = append(a.workerList, w)
+	}
+	for message := range a.messageChan {
+		channel := <-a.workerPool
+		channel <- message
+	}
+}
+
+func (a *aggregation) Stop() {
+	for _, worker := range a.workerList {
+		worker.stop()
+	}
+}
+
+// AggregateMsgBy5Tuple gets 5-tuple info from records in message and stores in cache
+func (a *aggregation) AggregateMsgBy5Tuple(message *entities.Message) error {
+	addOriginalExporterInfo(message)
```

Comment:
Should we move addOriginalExporterInfo() after line 77 since we skip Template Sets?

Reply:
For template records, currently we keep the `originalExporterIPv4Address` and `originalObservationDomainId` in case we want to re-use the elements from template record when exporting message from flow aggregator.